tools/evmc/main.cpp

```diff
@@ -0,0 +1,128 @@
+
+inline std::string hex(uint8_t b) noexcept
+{
+    static constexpr auto hex_chars = "0123456789abcdef";
+    return {hex_chars[b >> 4], hex_chars[b & 0xf]};
+}
+
+bytes from_hex(const std::string& hex)
+{
+    if (hex.length() % 2 == 1)
+        throw std::length_error{"the length of the input is odd"};
+
+    bytes bs;
+    int b = 0;
+    for (size_t i = 0; i < hex.size(); ++i)
+    {
+        auto h = hex[i];
+        int v;
+        if (h >= '0' && h <= '9')
+            v = h - '0';
+        else if (h >= 'a' && h <= 'f')
+            v = h - 'a' + 10;
+        else if (h >= 'A' && h <= 'F')
+            v = h - 'A' + 10;
+        else
+            throw std::out_of_range{"not a hex digit"};
+
+        if (i % 2 == 0)
+            b = v << 4;
+        else
+            bs.push_back(static_cast<uint8_t>(b | v));
+    }
+    return bs;
+}
+
+std::string hex(const bytes& bs)  // FIXME: Use bytes_view.
+{
+    std::string str;
+    str.reserve(bs.size() * 2);
+    for (auto b : bs)
+        str += hex(b);
+    return str;
+}
+}  // namespace evmc
+
+int main(int argc, const char** argv)
+{
+    CLI::App app{"EVMC tool"};
+    const auto& version_flag = *app.add_flag("--version", "Print version information");
+
+    std::string vm_config;
+    std::string code_hex;
+    auto& run_cmd = *app.add_subcommand("run", "Execute EVM bytecode");
+    run_cmd.add_option("--vm", vm_config)->required();
+    run_cmd.add_option("code", code_hex)->required();
+
+    try
+    {
+        app.parse(argc, argv);
+
+        // Handle the --version flag first and exit when present.
+        if (version_flag)
+        {
+            std::cout << "EVMC tool " PROJECT_VERSION "\n";
+            return 0;
+        }
+
+        if (run_cmd)
+        {
+            auto code = evmc::from_hex(code_hex);
+
+            std::cout << "Running " << vm_config << "\n";
+            evmc_loader_error_code ec;
+            auto vm = evmc::VM{evmc_load_and_configure(vm_config.c_str(), &ec)};
+            if (ec != EVMC_LOADER_SUCCESS)
+            {
+                const auto error = evmc_last_error_msg();
+                if (error != nullptr)
+                    std::cerr << error << "\n";
+                else
+                    std::cerr << "Loading error " << ec << "\n";
+                return static_cast<int>(ec);
+            }
+
+            evmc::MockedHost host;
+            evmc_message msg{};
+            msg.gas = 10000000;
+
+            auto result = vm.execute(host, EVMC_ISTANBUL, msg, code.data(), code.size());
+
+            auto output = evmc::bytes{result.output_data, result.output_data + result.output_size};
+            auto output_hex = evmc::hex(output);
+
+            auto gas_used = msg.gas - result.gas_left;
+
+            std::cout << "Output: " << output_hex << "\n"
+                      << "Gas used: " << gas_used << "\n";
+
+            return 0;
+        }
+
+        return 0;
+    }
+    catch (const CLI::ParseError& e)
```

Comment:
This should be better only around `app.parse` call, nothing else throws `ParseError`

Reply:
It can be, but isn't the point of exceptions to have single place for handling them and having the main code flow uninterrupted?

Reply:
Ok true, maybe my intuition was wrong